File: codes/go/gemini/minimal/task_4/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
)

func allocateResources(patients [][]int, resources []map[string]interface{}, responders int, totalTime int) map[string]interface{} {
	assignments := []map[string]interface{}{}
	survivors := 0
	totalSurvivalProbability := 0.0

	patientCount := len(patients)
	resourceCount := len(resources)

	// dp[i][j] stores the maximum survival probability using the first i patients and j time units
	dp := make([][]float64, patientCount+1)
	for i := range dp {
		dp[i] = make([]float64, totalTime+1)
	}

	// Keep track of assignments made during DP calculation
	assignmentTracker := make([][][]map[string]interface{}, patientCount+1)
	for i := range assignmentTracker {
		assignmentTracker[i] = make([][]map[string]interface{}, totalTime+1)
		for j := range assignmentTracker[i] {
			assignmentTracker[i][j] = []map[string]interface{}{}
		}
	}

	// Iterate through each patient and consider all possible resource assignments
	for i := 1; i <= patientCount; i++ {
		for j := 0; j <= totalTime; j++ {
			dp[i][j] = dp[i-1][j] // Initialize with the probability without treating the current patient
			assignmentTracker[i][j] = assignmentTracker[i-1][j]

			// Consider all possible resource assignments for the current patient
			for r := 0; r < resourceCount; r++ {
				resourceTime := int(resources[r]["time"].(float64))
				resourceBoost := int(resources[r]["boost"].(float64))

				// Check if there is enough time to apply the resource
				if resourceTime <= j {
					// Calculate the survival probability with the resource applied
					treatmentTime := j - resourceTime
					baseProbability := float64(patients[i-1][treatmentTime])
					boostedProbability := math.Min(100.0, baseProbability+float64(resourceBoost))

					// Check if applying the resource increases the survival probability
					if dp[i-1][j-resourceTime]+boostedProbability > dp[i][j] {
						dp[i][j] = dp[i-1][j-resourceTime] + boostedProbability

						// Update the assignment tracker
						currentAssignments := make([]map[string]interface{}, len(assignmentTracker[i-1][j-resourceTime]))
						copy(currentAssignments, assignmentTracker[i-1][j-resourceTime])

						currentAssignments = append(currentAssignments, map[string]interface{}{
							"patient":  i - 1,
							"resource": r,
							"time":     treatmentTime,
						})
						assignmentTracker[i][j] = currentAssignments
					}
				}
			}
		}
	}

	// Extract the optimal assignments
	assignments = assignmentTracker[patientCount][totalTime]

	// Calculate the number of survivors and total survival probability
	survivorMap := make(map[int]bool)
	for _, assignment := range assignments {
		patientIndex := int(assignment["patient"].(int))
		resourceIndex := int(assignment["resource"].(int))
		treatmentTime := int(assignment["time"].(int))

		baseProbability := float64(patients[patientIndex][treatmentTime])
		resourceBoost := int(resources[resourceIndex]["boost"].(float64))
		boostedProbability := math.Min(100.0, baseProbability+float64(resourceBoost))

		totalSurvivalProbability += boostedProbability
		survivorMap[patientIndex] = true
	}

	survivors = len(survivorMap)

	result := map[string]interface{}{
		"assignments":               assignments,
		"survivors":                 survivors,
		"total_survival_probability": totalSurvivalProbability,
	}

	return result
}

func main() {
	var input map[string]interface{}
	decoder := json.NewDecoder(os.Stdin)
	err := decoder.Decode(&input)
	if err != nil {
		fmt.Println("Error decoding JSON:", err)
		return
	}

	patients := input["patients"].([][]int)
	resources := input["resources"].([]map[string]interface{})
	responders := int(input["responders"].(float64))
	totalTime := int(input["total_time"].(float64))

	result := allocateResources(patients, resources, responders, totalTime)

	jsonResult, err := json.Marshal(result)
	if err != nil {
		fmt.Println("Error encoding JSON:", err)
		return
	}

	fmt.Println(string(jsonResult))
}

File: codes/javascript/gemini/minimal/task_4/Solution.js
function allocateResources(patients, resources, responders, totalTime) {
  const assignments = [];
  let survivors = 0;
  let total_survival_probability = 0;

  if (!patients || patients.length === 0 || !resources || resources.length === 0 || !responders || responders <= 0 || !totalTime || totalTime <= 0) {
    return { assignments: [], survivors: 0, total_survival_probability: 0 };
  }

  const patientCount = patients.length;
  const resourceCount = resources.length;

  const possibleAssignments = [];

  for (let i = 0; i < patientCount; i++) {
    for (let j = 0; j < resourceCount; j++) {
      for (let t = 0; t <= totalTime - resources[j].time; t++) {
        const baseProbability = patients[i][t];
        const boostedProbability = Math.min(100, baseProbability + resources[j].boost);

        let finalProbability = boostedProbability;
        
        possibleAssignments.push({
          patient: i,
          resource: j,
          time: t,
          probability: finalProbability / 100, 
        });
      }
    }
  }
  
  possibleAssignments.sort((a, b) => b.probability - a.probability);

  const responderAvailability = new Array(responders).fill(0);
  const assignedPatients = new Set();

  for (const assignment of possibleAssignments) {
    const patient = assignment.patient;
    const resource = assignment.resource;
    const time = assignment.time;
    const probability = assignment.probability;

    if (assignedPatients.has(patient)) {
      continue;
    }

    let availableResponderIndex = -1;
    for (let i = 0; i < responders; i++) {
      if (responderAvailability[i] <= time) {
        availableResponderIndex = i;
        break;
      }
    }

    if (availableResponderIndex !== -1) {
      assignments.push({
        patient: patient,
        resource: resource,
        time: time,
      });
      responderAvailability[availableResponderIndex] = time + resources[resource].time;
      assignedPatients.add(patient);
    }
  }

  for (let i = 0; i < patients.length; i++) {
    let survivalProbability = 0;
    let assignedTime = -1;
    let assignedResourceIndex = -1;

    for (const assignment of assignments) {
      if (assignment.patient === i) {
        assignedTime = assignment.time;
        assignedResourceIndex = assignment.resource;
        break;
      }
    }

    if (assignedTime !== -1) {
        const baseProbability = patients[i][assignedTime];
        survivalProbability = Math.min(100, baseProbability + resources[assignedResourceIndex].boost);
        survivors++;
        total_survival_probability += survivalProbability;
      
    }
  }

  return {
    assignments: assignments,
    survivors: survivors,
    total_survival_probability: total_survival_probability,
  };
}

if (typeof process !== 'undefined' && process.argv) {
  let inputData = '';

  process.stdin.on('data', (chunk) => {
    inputData += chunk;
  });

  process.stdin.on('end', () => {
    try {
      const parsedInput = JSON.parse(inputData);
      const result = allocateResources(parsedInput.patients, parsedInput.resources, parsedInput.responders, parsedInput.total_time);
      console.log(JSON.stringify(result));
    } catch (error) {
      console.error('Error parsing JSON input:', error);
      console.error('Input data:', inputData);
    }
  });
}